lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
+
+    /// Create an UnboundImport for a user-written import declaration.
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
+
+    /// Make sure the module actually loaded, and diagnose if it didn't.
+    bool checkModuleLoaded(ModuleDecl *M, SourceFile &SF);
+
+    /// Find the top-level module for this module. If \p M is not a submodule,
+    /// returns \p M. If it is a submodule, returns either the parent module of
+    /// \p M or \c nullptr if the current module is the parent (which can happen
+    /// in a mixed-language framework).
+    NullablePtr<ModuleDecl> getTopLevelModule(ModuleDecl *M, SourceFile &SF);
+
+    /// Diagnose any errors concerning the \c @_exported, \c @_implementationOnly,
+    /// \c @testable, or \c @_private attributes, including a
+    /// non-implementation-only import of a fragile library from a resilient one.
+    void validateOptions(NullablePtr<ModuleDecl> topLevelModule, SourceFile &SF);
+
+    /// Create an \c ImportedModuleDesc from the information in this
+    /// UnboundImport.
+    ImportedModuleDesc makeDesc(ModuleDecl *module) const {
+      return ImportedModuleDesc({ declPath, module }, options,
+                                privateImportFileName);
+    }
 
-namespace {  
-  class NameBinder {    
-  public:
+  private:
+    void validatePrivate(ModuleDecl *topLevelModule);
+    void validateImplementationOnly(ASTContext &ctx);
+    void validateTestable(ModuleDecl *topLevelModule);
+    void validateResilience(NullablePtr<ModuleDecl> topLevelModule,
+                            SourceFile &SF);
+
+    /// Diagnoses an inability to import \p modulePath in this situation and, if
+    /// \p attrs is provided and has an \p attrKind, invalidates the attribute and
+    /// offers a fix-it to remove it.
+    void diagnoseInvalidAttr(DeclAttrKind attrKind, DiagnosticEngine &diags,
+                             Diag<Identifier> diagID);
+  };
+
+  /// Represents an import whose options have been checked and module has been
+  /// loaded, but its scope (if any) has not been validated.
+  struct BoundImport {
+    UnboundImport unbound;
+    ImportedModuleDesc desc;
+    ModuleDecl *module;
+    bool needsScopeValidation = false;
+
+    BoundImport(UnboundImport unbound, ImportedModuleDesc desc,
+                ModuleDecl *module, bool needsScopeValidation);
+
+    /// Validate the scope of the import, if needed.
+    void validateScope(SourceFile &SF);
+  };
+
+  class NameBinder : public DeclVisitor<NameBinder> {
     SourceFile &SF;
-    ASTContext &Context;
+    ASTContext &ctx;
```

Comment:
Accidental rename?

Reply:
No, intentional; I'm using `ctx` in new code and didn't want it to be inconsistent.